Abstract out AddressLayout and AddressSpace function.

jonseymour opened the pull request:
The end goal of these changes is to support non-SunSpec address
spaces with the same API. The reason is that in any practical SunSpec
work, it will be necessary to access parts of the address space
that is not layed out with SunSpec conventions - usually legacy, vendor
specfic address maps that have not been standardised by SunSpec and
have not been developed in accordance with SunSpec design guidelines
(e.g. they are not self-describing).

Now, while we could create a whole different API for this purpose, but this API
already does a vast majority of the heavy lifting with regard to meta-data
management and data marshalling, so it seems sensible to enable re-use of this work.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour commented:
Hello @crabmusket, Happy New Year to you.

These changes allows the main API to be used with ModBus address spaces that have non-SunSpec compliant layouts. With the changes we can re-use the SMDX to describe blocks of memory, but there is no requirement that the blocks be contiguous or be discoverable with model + length encodings (as per SunSpec).

The intent is to allow one API to be used to access devices that have a SunSpec compliant address space and also a non-SunSpec compliant address space.

At this point in time, I have only implemented support for the SunSpecLayout, but intend to add support for the so-called "RawLayout" which allows devices to be comprised of blocks in arbitrary regions of the address space.

jonseymour committed:
sunspec.Block: add DoScaleFactorsFirst

Sometimes we need to iterate in scale factor first order because we need
the order of writes to be scale factor then value.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

crabmusket commented:
Same to you :). This seems like a good idea, and important in practise. Merge when ready!

jonseymour committed:
Add additional logging.

jonseymour committed:
Fix support for the 304 model so that nr is the number of additional repeats, rather than the total number of repeats.

(This probably doesn't handle 304 models with zero length, but a change
can be made separately as and when required for this case).

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Abstract out AddressLayout and AddressSpace function.

The purpose of these changes is to support non-SunSpec address
spaces with the same API. The reason is that in any practical SunSpec
work, it will be necessary to access parts of the address space
that is not layed out with SunSpec conventions - usually legacy, vendor
specfic address maps that have not been standardised by SunSpec and
have not been developed in accordance with SunSpec design guidelines
(e.g. they are not self-describing).

Now, we could create a whole different API for this purpose, but this API
already does a vast majority of the heavy lifting with regard to meta-data
management and data marshalling, so it seems sensible to enable re-use of this work.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>
